pkg/devcontainer/config/userenvprobe.go

```diff
@@ -0,0 +1,181 @@
+	if err != nil {
+		log.Warnf("Get user env probe: %v", err)
+		log.Warnf("Falling back to default user env probe: %s", DefaultUserEnvProbe)
+		userEnvProbe = DefaultUserEnvProbe
+	}
+	if userEnvProbe == NoneProbe {
+		return map[string]string{}, nil
+	}
+
+	preferredShell, err := shell.GetShell(userName)
+	if err != nil {
+		return nil, fmt.Errorf("find shell for user %s: %w", userName, err)
+	}
+
+	log.Debugf("running user env probe with shell \"%s\", probe \"%s\", user \"%s\" and command \"%s\"",
+		strings.Join(preferredShell, " "), string(userEnvProbe), userName, "cat /proc/self/environ")
+
+	probedEnv, err := doProbe(ctx, userEnvProbe, preferredShell, userName, "cat /proc/self/environ", '\x00', log)
+	if err != nil {
+		log.Debugf("running user env probe with shell \"%s\", probe \"%s\", user \"%s\" and command \"%s\"",
+			strings.Join(preferredShell, " "), string(userEnvProbe), userName, "printenv")
+
+		newProbedEnv, newErr := doProbe(ctx, userEnvProbe, preferredShell, userName, "printenv", '\n', log)
+		if newErr != nil {
+			log.Warnf("failed to probe user environment variables: %v, %v", err, newErr)
+		} else {
+			probedEnv = newProbedEnv
+		}
+	}
+	if probedEnv == nil {
+		probedEnv = map[string]string{}
+	}
+
+	return probedEnv, nil
+}
+
+func getShellArgs(userEnvProbe UserEnvProbe, user, command string) []string {
+	args := []string{}
+	switch userEnvProbe {
+	case LoginInteractiveShellProbe:
+		args = append(args, "-lic")
+	case LoginShellProbe:
+		args = append(args, "-lc")
+	case InteractiveShellProbe:
+		args = append(args, "-ic")
+	// shouldn't happen, added just for linting
+	case NoneProbe:
+		args = append(args, "-c")
+	default:
+		args = append(args, "-c")
+	}
+	args = append(args, command)
+
+	return args
+}
+
+func doProbe(ctx context.Context, userEnvProbe UserEnvProbe, preferredShell []string, userName string, probeCmd string, sep byte, log log.Logger) (map[string]string, error) {
+	args := preferredShell
+	args = append(args, getShellArgs(userEnvProbe, userName, probeCmd)...)
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+	cmd := exec.CommandContext(timeoutCtx, args[0], args[1:]...)
+
+	err := PrepareProbe(cmd, userName)
+	if err != nil {
+		return nil, fmt.Errorf("prepare probe: %w", err)
+	}
+
+	out, err := cmd.Output()
+	if err != nil {
+		return nil, fmt.Errorf("probe user env: %w", err)
+	}
+
+	scanner := bufio.NewScanner(bytes.NewBuffer(out))
+	scanner.Split(splitBySeparator(sep))
+
+	retEnv := map[string]string{}
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		tokens := strings.Split(line, "=")
+		if len(tokens) == 1 {
+			log.Debugf("failed to split env var: %s", line)
+			continue
+		}
+		retEnv[tokens[0]] = tokens[1]
+	}
+	if scanner.Err() != nil {
+		return nil, fmt.Errorf("scan shell output: %w", err)
+	}
+	delete(retEnv, "PWD")
+
+	return retEnv, nil
+}
+
+func patchEnvVars(env []string, patches map[string]string) []string {
```

Comment:
similar, also feel free to ignore: since this is used only by platform-specific implementation I'd either move it closer to this implementation or to the bottom to this file with some comment about it's usage/purpose

Reply:
sure thing, done